### PR TITLE
Load both ~/.bash_history and ~/.black-screen/history

### PR DESCRIPTION
**Pull request by drew-gross**
Due to some bugs in black-screen I switch between iTerm2 and black-screen. This change reads the command history from both history files, putting the one that was written to most recently first.

### CODE REVIEW

**Comment by ekaragodin** [comment]
It would be great to support zsh_history

**Comment by shockone** [comment]
Hey, @drew-gross. It might be a good short-term fix, but I'd rather not do it for two reasons:

* We plan to extend the history format to include start time, end time, session id, PWD and maybe something else. It'll make them incompatible.
* We should really make it stable instead of fixing the symptoms.

**Comment by shockone** [comment]
Although, we might leave it for now.

File: src/utils/Common.ts [hunk 1]
```diff
@@ -350,4 +350,6 @@ export function assign<A extends B, B extends Object>(source: A, assignments: B)
 const baseConfigDirectory = Path.join(homeDirectory, ".black-screen");
 export const presentWorkingDirectoryFilePath = Path.join(baseConfigDirectory, "presentWorkingDirectory");
 export const historyFilePath = Path.join(baseConfigDirectory, "history");
+// We can't use HISTFILE to get the path, so just check the default location
+export const bashHistoryFilePath = Path.join(homeDirectory, ".bash_history");
```
Comment (shockone) [inline src/utils/Common.ts]: There is a better place for that. https://github.com/shockone/black-screen/blob/master/src/utils/Shell.ts. You can also move deserialization logic there, if it differs across shells.

File: src/shell/Session.ts [hunk 2]
```diff
@@ -112,7 +113,32 @@ export class Session extends EmitterWithUniqueID {
 
     private deserialize(): void {
         this.directory = this.readSerialized(presentWorkingDirectoryFilePath, homeDirectory);
-        History.deserialize(this.readSerialized(historyFilePath, []));
+        let bashHistoryModifiedTime = new Date(0);
```
Comment (shockone) [inline src/shell/Session.ts]: I'd say this code belongs to `History.deserialize`.

File: src/shell/Session.ts [hunk 3]
```diff
@@ -112,7 +113,32 @@ export class Session extends EmitterWithUniqueID {
...
+            blackScreenHistory = this.readSerialized(historyFilePath, []);
+        } catch (error) {
+            // File was missing or something. Ignore.
+        }
+
+        let mergedHistory: string[] = [];
+        if (blackScreenHistoryModifiedTime > bashHistoryModifiedTime) {
+            mergedHistory = [...bashHistory, ...blackScreenHistory];
+        } else {
+            mergedHistory = [...blackScreenHistory, ...bashHistory];
```
Comment (shockone) [inline src/shell/Session.ts]: Would it be possible to interleave the records based on time? It's not trivial to do, so I'm OK with a "no".

File: src/shell/Session.ts [hunk 4]
```diff
@@ -112,7 +113,32 @@ export class Session extends EmitterWithUniqueID {
 
     private deserialize(): void {
         this.directory = this.readSerialized(presentWorkingDirectoryFilePath, homeDirectory);
-        History.deserialize(this.readSerialized(historyFilePath, []));
+        let bashHistoryModifiedTime = new Date(0);
+        let bashHistory: string[] = [];
+        try {
+            bashHistoryModifiedTime = statSync(bashHistoryFilePath).mtime;
+            bashHistory = readFileSync(bashHistoryFilePath).toString().trim().split("\n").reverse();
```
Comment (shockone) [inline src/shell/Session.ts]: `"\n"` should be `os.EOL`.

File: src/shell/Session.ts (hunk 3 above)
Comment (drew-gross) [inline src/shell/Session.ts]: As far as I can tell, the time is not stored in `.bash_history`.

**Comment by drew-gross** [comment]
I hope we can also make this stable, but even once it is, I think loading their original shell history will still be nice, especially for people switching to black-screen for the first time.

File: src/shell/Session.ts (hunk 3 above)
Comment (shockone) [inline src/shell/Session.ts]: Actually, there is. Or rather you can configure the shell to have it. So, if not to support it, we need to at least verify the format we read is expected. The keyword to find more is `HISTTIMEFORMAT`.

**Comment by drew-gross** [comment]
OK, I made the suggested changes.

File: src/utils/Shell.ts [hunk 5]
```diff
@@ -78,4 +113,8 @@ const {shell} = new class {
     };
 };
 
+export function loadAllHistories(): { lastModified: Date, commands: string[] }[] {
```
Comment (shockone) [inline src/utils/Shell.ts]: You can safely assume that a person only wants to read history from his login shell, which is `loginShell`. Also, it'll be consistent with alias loading behaviour. So, please remove this function and use `loginShell. loadHistory()` instead.

File: src/utils/Shell.ts [hunk 6]
```diff
@@ -28,6 +34,27 @@ class Bash extends Shell {
     get preCommandModifiers(): string[] {
         return [];
     }
+
+    loadHistory(): { lastModified: Date, commands: string[] } {
+        if (process.env.HISTTIMEFORMAT) {
```
Comment (shockone) [inline src/utils/Shell.ts]: This check is pretty much useless: Black Screen will only pick up shell's environment if it's started from a shell session. But usually it's started as a standalone application.

**Comment by drew-gross** [comment]
Done.

**Pull request merged**